File: Content/Code/deque-fixed-circular-array/cpp/fixed_array_deque_exercise.cpp
#include <iostream>

using namespace std;

class IntDeque {
private:
int* A;
int capacity;
int front;
int count;

int backIndex() {
    // TODO
    return -1;
}

public:
IntDeque(int cap) {
    capacity = cap;
    A = new int[capacity];
    front = 0;
    count = 0;
}

~IntDeque() {
    delete[] A;
}

bool isEmpty() {
    // TODO
    return false;
}

bool isFull() {
    // TODO
    return false;
}

int size() {
    // TODO
    return -1;
}

void clear() {
    // TODO
}

int peekFront() {
    // TODO
    return -1;
}

int peekBack() {
    // TODO
    return -1;
}

bool addFront(int value) {
    // TODO
    return false;
}

bool addBack(int value) {
    // TODO
    return false;
}

int removeFront() {
    // TODO
    return -1;
}

int removeBack() {
    // TODO
    return -1;
}

};

void check(int actual, int expected) {
    if (actual == expected) {
        cout << "pass" << endl;
    } else {
        cout << "fail: expected " << expected << " but got " << actual << endl;
    }
}

void check(bool actual, bool expected) {
    if (actual == expected) {
        cout << "pass" << endl;
    } else {
        cout << "fail: expected " << expected << " but got " << actual << endl;
    }
}

void testDeque() {
IntDeque deque(5);

check(deque.isEmpty(), true);
check(deque.isFull(), false);
check(deque.size(), 0);
check(deque.peekFront(), -1);
check(deque.peekBack(), -1);
check(deque.removeFront(), -1);
check(deque.removeBack(), -1);

check(deque.addBack(4), true);       // [4]
check(deque.peekFront(), 4);
check(deque.peekBack(), 4);
check(deque.size(), 1);

check(deque.addBack(7), true);       // [4, 7]
check(deque.addFront(2), true);      // [2, 4, 7]
check(deque.addBack(9), true);       // [2, 4, 7, 9]

check(deque.size(), 4);
check(deque.isEmpty(), false);
check(deque.peekFront(), 2);
check(deque.peekBack(), 9);

check(deque.removeFront(), 2);       // [4, 7, 9]
check(deque.peekFront(), 4);
check(deque.removeBack(), 9);        // [4, 7]
check(deque.peekBack(), 7);
check(deque.size(), 2);

check(deque.addFront(1), true);      // [1, 4, 7]
check(deque.addFront(0), true);      // [0, 1, 4, 7]
check(deque.addBack(11), true);      // [0, 1, 4, 7, 11]
check(deque.isFull(), true);
check(deque.size(), 5);
check(deque.peekFront(), 0);
check(deque.peekBack(), 11);

check(deque.addBack(13), false);     // full
check(deque.addFront(13), false);    // full
check(deque.size(), 5);
check(deque.peekFront(), 0);
check(deque.peekBack(), 11);

check(deque.removeFront(), 0);       // [1, 4, 7, 11]
check(deque.removeFront(), 1);       // [4, 7, 11]
check(deque.addBack(13), true);      // [4, 7, 11, 13]
check(deque.addBack(15), true);      // [4, 7, 11, 13, 15]
check(deque.isFull(), true);
check(deque.peekFront(), 4);
check(deque.peekBack(), 15);

check(deque.removeBack(), 15);       // [4, 7, 11, 13]
check(deque.removeBack(), 13);       // [4, 7, 11]
check(deque.addFront(3), true);      // [3, 4, 7, 11]
check(deque.addFront(2), true);      // [2, 3, 4, 7, 11]
check(deque.peekFront(), 2);
check(deque.peekBack(), 11);

check(deque.removeFront(), 2);       // [3, 4, 7, 11]
check(deque.removeBack(), 11);       // [3, 4, 7]
check(deque.removeFront(), 3);       // [4, 7]
check(deque.removeBack(), 7);        // [4]
check(deque.removeBack(), 4);        // []

check(deque.isEmpty(), true);
check(deque.size(), 0);
check(deque.peekFront(), -1);
check(deque.peekBack(), -1);
check(deque.removeFront(), -1);
check(deque.removeBack(), -1);

check(deque.addFront(6), true);      // [6]
check(deque.addBack(8), true);       // [6, 8]
deque.clear();                       // []
check(deque.isEmpty(), true);
check(deque.size(), 0);
check(deque.peekFront(), -1);

}

int main() {
  testDeque();
  return 0;
}

File: Content/Code/deque-fixed-circular-array/python/fixed_array_deque_exercise.py
import inspect


class IntDeque:
    def __init__(self, capacity):
        self.A = [0] * capacity
        self.front = 0
        self.count = 0

    def is_empty(self):
        # TODO
        return False

    def is_full(self):
        # TODO
        return False

    def size(self):
        # TODO
        return -1

    def clear(self):
        # TODO
        pass

    def back_index(self):
        # TODO
        return -1

    def peek_front(self):
        # TODO
        return -1

    def peek_back(self):
        # TODO
        return -1

    def add_front(self, value):
        # TODO
        return False

    def add_back(self, value):
        # TODO
        return False

    def remove_front(self):
        # TODO
        return -1

    def remove_back(self):
        # TODO
        return -1


def check(actual, expected):
    line = inspect.currentframe().f_back.f_lineno
    if actual == expected:
        print(f"PASS at test line {line}: got {actual!r}")
    else:
        print(f"FAIL at test line {line}: expected {expected!r} but got {actual!r}")


def test_deque():
    deque = IntDeque(5)

    check(deque.is_empty(), True)
    check(deque.is_full(), False)
    check(deque.size(), 0)
    check(deque.peek_front(), -1)

    large = IntDeque(257)
    large_ok = True
    for round_number in range(20):
        for i in range(257):
            large_ok = large.add_back(round_number * 257 + i) and large_ok
        large_ok = large.is_full() and not large.add_front(-1) and large_ok
        for i in range(257):
            large_ok = large.remove_front() == round_number * 257 + i and large_ok
    check(large_ok and large.is_empty(), True)
    check(deque.peek_back(), -1)
    check(deque.remove_front(), -1)
    check(deque.remove_back(), -1)

    check(deque.add_back(4), True)       # [4]
    check(deque.peek_front(), 4)
    check(deque.peek_back(), 4)
    check(deque.size(), 1)

    check(deque.add_back(7), True)       # [4, 7]
    check(deque.add_front(2), True)      # [2, 4, 7]
    check(deque.add_back(9), True)       # [2, 4, 7, 9]

    check(deque.size(), 4)
    check(deque.is_empty(), False)
    check(deque.peek_front(), 2)
    check(deque.peek_back(), 9)

    check(deque.remove_front(), 2)       # [4, 7, 9]
    check(deque.peek_front(), 4)
    check(deque.remove_back(), 9)        # [4, 7]
    check(deque.peek_back(), 7)
    check(deque.size(), 2)

    check(deque.add_front(1), True)      # [1, 4, 7]
    check(deque.add_front(0), True)      # [0, 1, 4, 7]
    check(deque.add_back(11), True)      # [0, 1, 4, 7, 11]
    check(deque.is_full(), True)
    check(deque.size(), 5)
    check(deque.peek_front(), 0)
    check(deque.peek_back(), 11)

    check(deque.add_back(13), False)     # full
    check(deque.add_front(13), False)    # full
    check(deque.size(), 5)
    check(deque.peek_front(), 0)
    check(deque.peek_back(), 11)

    check(deque.remove_front(), 0)       # [1, 4, 7, 11]
    check(deque.remove_front(), 1)       # [4, 7, 11]
    check(deque.add_back(13), True)      # [4, 7, 11, 13]
    check(deque.add_back(15), True)      # [4, 7, 11, 13, 15]
    check(deque.is_full(), True)
    check(deque.peek_front(), 4)
    check(deque.peek_back(), 15)

    check(deque.remove_back(), 15)       # [4, 7, 11, 13]
    check(deque.remove_back(), 13)       # [4, 7, 11]
    check(deque.add_front(3), True)      # [3, 4, 7, 11]
    check(deque.add_front(2), True)      # [2, 3, 4, 7, 11]
    check(deque.peek_front(), 2)
    check(deque.peek_back(), 11)

    check(deque.remove_front(), 2)       # [3, 4, 7, 11]
    check(deque.remove_back(), 11)       # [3, 4, 7]
    check(deque.remove_front(), 3)       # [4, 7]
    check(deque.remove_back(), 7)        # [4]
    check(deque.remove_back(), 4)        # []

    check(deque.is_empty(), True)
    check(deque.size(), 0)
    check(deque.peek_front(), -1)
    check(deque.peek_back(), -1)
    check(deque.remove_front(), -1)
    check(deque.remove_back(), -1)

    check(deque.add_front(6), True)      # [6]
    check(deque.add_back(8), True)       # [6, 8]
    deque.clear()                        # []
    check(deque.is_empty(), True)
    check(deque.size(), 0)
    check(deque.peek_front(), -1)


test_deque()
